### DISCUSSION TITLE
Process not exiting after `queue.add()`

### DISCUSSION THREAD

**Discussion by SevenOutman** [OptimalBits/bull#2107] [Q&A]
I have a CLI script in which I need to send an email. I choose to use queues to defer the email sending job so that CLI user don't have to wait for it.

Now I have a script for register queues' process functions that will be run as daemon. And I have a separate CLI script in which I add jobs to queues.

queue.js
```js
const { mailQueue } = require('./queues')
const { sendMail } = require('./mail')

mailQueue.process(async (job) => {
  return sendMail(job.data)
});
```

cli.js
```js
const { mailQueue } = require('./queues')

(async () => {
  await mailQueue.add({
    receiver,
    subject,
    text
  });
  console.debug('Job added')
})();
```

But when I run the CLI script, the process is not exiting after adding the job, even after the email is sent. Can't figure out why.

**Comment by devopg** [comment]
up same problem

**Comment by manast** [comment]
You need to close the queue:
```ts
await mailQueue.close()
```

**Reply by devopg** [reply]
how isit work ? Internal Bull create subprocess which dont allow close parrent ? I think parent nodejs process has proprity and then finish terminate all thread/subproccess

**Reply by SevenOutman** [reply]
I see. It works but I think the `.close()` API is not as expressive. I assume \*closing the queue\* means shutting down the queue and jobs in the queue would be lost then. If what the method does is releasing the handle that attached to the queue, I'd suggest a `.disconnect()` naming.🤔

**Reply by manast** [reply]
@SevenOutman No. Closing the queue does not imply any jobs to be "lost". ```queue.close()``` will wait for current jobs to be processed, whereas ```queue.close(true)```. will force closing without waiting, this is  documented here: https://github.com/OptimalBits/bull/blob/develop/REFERENCE.md#queueclose

**Reply by SevenOutman** [reply]
> queue.close() will wait for current jobs to be processed

This said, does it still hang my process after calling `.close()`?

**Reply by manast** [reply]
That depends on whether your jobs finish or not...

**Reply by SevenOutman** [reply]
If I'm getting you correctly, that sounds pretty weird to me 🤨. The purpose of using queues is to allow user don't have to wait until time-consuming tasks are finished. Instead they can get response right after tasks are just committed. If I can't manage to exit after adding the job but have to wait until it's finished, what's the point in using a queue then?

**Reply by manast** [reply]
I think you got things mixed up here. If you have a Queue instance that is processing jobs and you close that instance what do you expect to happen?

**Reply by manast** [reply]
nvm. In your case above you need to close the queue for node to exit.
```ts
const { mailQueue } = require('./queues')

(async () => {
  await mailQueue.add({
    receiver,
    subject,
    text
  });
  console.debug('Job added')
await mailQueue.close()
})();
```

And if it still does not exit then you use this:

https://www.npmjs.com/package/why-is-node-running

**Reply by SevenOutman** [reply]
> If you have a Queue instance that is processing jobs and you close that instance what do you expect to happen?

I have two separate scripts for this queue workflow. One running as a daemon in which `queue.process` logic is defined. The other is a single-run script where user commit a job with `queue.add` call and exit right then, not having to wait for that job to finish. The problem I describe happens in this single-run process.

> nvm. In your case above you need to close the queue for node to exit.

Does this `.close()` call hang the process until the job just added is finished?
